build.assets/tooling/cmd/goderive/plugin/deepcopy/deepcopy.go

```diff
@@ -0,0 +1,423 @@
+//   - pointer
+//   - slice
+//   - map
+//
+// , otherwise we are not able to modify the input parameter and then what are you really copying,
+// but as we go deeper we support most types.
+//
+// Supported types:
+//   - basic types
+//   - named structs
+//   - slices
+//   - maps
+//   - pointers to these types
+//   - private fields of structs in external packages (using reflect and unsafe)
+//   - and many more
+//
+// Unsupported types:
+//   - chan
+//   - interface
+//   - function
+//   - unnamed structs, which are not comparable with the == operator
+//
+// Example output can be found here:
+// https://github.com/awalterschulze/goderive/tree/master/example/plugin/deepcopy
+//
+// This plugin has been tested thoroughly.
+package deepcopy
+
+import (
+	"fmt"
+	"go/types"
+	"strings"
+
+	"github.com/awalterschulze/goderive/derive"
+)
+
+// NewPlugin creates a new deepcopy plugin.
+// This function returns the plugin name, default prefix and a constructor for the deepcopy code generator.
+func NewPlugin() derive.Plugin {
+	return derive.NewPlugin("deepcopy", "deriveDeepCopy", New)
+}
+
+// New is a constructor for the deepcopy code generator.
+// This generator should be reconstructed for each package.
+func New(typesMap derive.TypesMap, p derive.Printer, deps map[string]derive.Dependency) derive.Generator {
+	return &gen{
+		TypesMap:   typesMap,
+		printer:    p,
+		bytesPkg:   p.NewImport("bytes", "bytes"),
+		reflectPkg: p.NewImport("reflect", "reflect"),
+		unsafePkg:  p.NewImport("unsafe", "unsafe"),
+	}
+}
+
+type gen struct {
+	derive.TypesMap
+	printer    derive.Printer
+	bytesPkg   derive.Import
+	reflectPkg derive.Import
+	unsafePkg  derive.Import
+}
+
+func (g *gen) Add(name string, typs []types.Type) (string, error) {
+	if len(typs) != 2 {
+		return "", fmt.Errorf("%s does not have two arguments", name)
+	}
+	if !types.Identical(typs[0], typs[1]) {
+		return "", fmt.Errorf("%s has two arguments, but they are of different types %s != %s",
+			name, g.TypeString(typs[0]), g.TypeString(typs[1]))
+	}
+	return g.SetFuncName(name, typs[0])
+}
+
+func (g *gen) Generate(typs []types.Type) error {
+	return g.genFunc(typs[0])
+}
+
+func (g *gen) genFunc(typ types.Type) error {
+	p := g.printer
+	g.Generating(typ)
+	typeStr := g.TypeString(typ)
+	p.P("")
+	p.P("// %s recursively copies the contents of src into dst.", g.GetFuncName(typ))
+	p.P("func %s(dst, src %s) {", g.GetFuncName(typ), typeStr)
+	p.In()
+	if err := g.genStatement(typ, "src", "dst"); err != nil {
+		return err
+	}
+	p.Out()
+	p.P("}")
+	return nil
+}
+
+func (g *gen) genStatement(typ types.Type, this, that string) error {
+	p := g.printer
+	if canCopy(typ) {
+		p.P("%s = %s", that, this)
+		return nil
+	}
+
+	if typ.String() == "*time.Time" {
+		p.P("if src.IsZero() {")
+		p.In()
+		p.P("return")
+		p.Out()
+		p.P("}")
+		p.P(fmt.Sprintf("*%s = time.Unix(0, %s.UnixNano()).In(%[2]s.Location())", that, this))
```

Comment:
Why this roundtrip through UnixNano (which narrows the time range)? Can't we just do `*%s = %s` or, better yet, add the type check to `canCopy` instead?

Reply:
yes, The `  p.P("*%s = *%s", that, this)` should be enough cc @tigrato